File: God Zuo/Python/class066 一维动态规划入门/My04.py
MOD = 10**9 + 7

class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        dp = [-1] * (n + 1)
        
        def f(i):
            if i == n:
                return 1
            if dp[i] != -1:
                return dp[i]
            if s[i] == '0':
                return 0
                
            ans = f(i + 1) * (9 if s[i] == '*' else 1)
            
            if i + 1 < n:
                if s[i] != '*':
                    if s[i + 1] != '*':
                        if int(s[i:i+2]) <= 26:
                            ans += f(i + 2)
                    else:
                        if s[i] == '1':
                            ans += f(i + 2) * 9
                        elif s[i] == '2':
                            ans += f(i + 2) * 6
                else:
                    if s[i + 1] != '*':
                        if s[i + 1] <= '6':
                            ans += f(i + 2) * 2
                        else:
                            ans += f(i + 2)
                    else:
                        ans += f(i + 2) * 15
                        
            ans %= MOD
            dp[i] = ans
            return ans
            
        return f(0)
    def numDecodings2(self, s: str) -> int:
        n = len(s)
        dp = [0] * (n + 1)
        dp[n] = 1
        
        for i in range(n - 1, -1, -1):
            if s[i] == '0':
                dp[i] = 0
            else:
                ans = dp[i + 1] * (9 if s[i] == '*' else 1)
                
                if i + 1 < n:
                    if s[i] != '*':
                        if s[i + 1] != '*':
                            if int(s[i:i+2]) <= 26:
                                ans += dp[i+2]
                        else:
                            if s[i] == '1':
                                ans += dp[i+2] * 9
                            elif s[i] == '2':
                                ans += dp[i+2] * 6
                    else:
                        if s[i + 1] != '*':
                            if s[i + 1] <= '6':
                                ans += dp[i+2] * 2
                            else:
                                ans += dp[i+2]
                        else:
                            ans += dp[i+2] * 15
                            
                dp[i] = ans % MOD
        return dp[0]

File: God Zuo/class069 三维动态规划/My04.cpp
// 测试链接 : https://leetcode.cn/problems/paths-in-matrix-whose-sum-is-divisible-by-k/
#include<bits/stdc++.h>
using namespace std;

class Solution{
public:
    const int MOD = 1000000007;
    vector<vector<vector<int>>> dp;
    int f(vector<vector<int>> &grid, int n, int m, int k, int i, int j, int r){
        if(i==n-1 && j==m-1)
            return grid[i][j] % k == r ? 1 : 0;
        if(dp[i][j][r]!=-1)
            return dp[i][j][r];
        int need = (k + r - (grid[i][j] % k)) % k;
        int ans = 0;
        if(i+1<n)
            ans = f(grid, n, m, k, i + 1, j, need)%MOD;
        if(j+1<m)
            ans = (ans+f(grid, n, m, k, i, j + 1, need))%MOD;
        dp[i][j][r] = ans;
        return ans;
    }

    int numberOfPaths(vector<vector<int>> &grid, int k){
        int n = grid.size();
        int m = grid[0].size();
        dp.assign(n, vector<vector<int>>(m, vector<int>(k, -1)));
        return f(grid, n, m, k, 0, 0, 0);
    }
};

class Solution
{
public:
    const int MOD = 1000000007;
    vector<vector<vector<int>>> dp;
    
    int numberOfPaths(vector<vector<int>> &grid, int k){
        int n = grid.size();
        int m = grid[0].size();
        dp.assign(n, vector<vector<int>>(m, vector<int>(k, 0)));
        dp[n - 1][m - 1][grid[n - 1][m - 1] % k] = 1;
        for (int i = n - 2; i >= 0;--i){
            for (int r = 0; r < k;++r){
                dp[i][m - 1][r] = dp[i + 1][m - 1][(k + r - grid[i][m - 1] % k) % k];
            }
        }
        for (int j = m - 2; j >= 0; j--)
        {
            for (int r = 0; r < k; r++)
            {
                dp[n - 1][j][r] = dp[n - 1][j + 1][(k + r - grid[n - 1][j] % k) % k];
            }
        }
        for (int i = n - 2; i >= 0;--i){
            for (int j = m - 2; j >= 0;--j){
                for (int r = 0; r < k;++r){
                    int need = (k + r - grid[i][j] % k) % k;
                    dp[i][j][r] = dp[i + 1][j][need];
                    dp[i][j][r] = (dp[i][j][r] + dp[i][j + 1][need]) % MOD;
                }
            }
        }
        return dp[0][0][0];
    }    
};